crates/ra_assists/src/handlers/extract_struct_from_enum_variant.rs

```diff
@@ -0,0 +1,327 @@
+    let target = variant.syntax().text_range();
+    return acc.add_in_multiple_files(
+        AssistId("extract_struct_from_enum_variant"),
+        "Extract struct from enum variant",
+        target,
+        |edit| {
+            let definition = Definition::ModuleDef(ModuleDef::EnumVariant(variant_hir));
+            let res = definition.find_usages(&ctx.db, None);
+            let start_offset = variant.parent_enum().syntax().text_range().start();
+            let mut visited_modules_set: FxHashSet<Module> = FxHashSet::default();
+            visited_modules_set.insert(current_module);
+            for reference in res {
+                let source_file = ctx.sema.parse(reference.file_range.file_id);
+                update_reference(
+                    ctx,
+                    edit,
+                    reference,
+                    &source_file,
+                    &current_module_def,
+                    &mut visited_modules_set,
+                );
+            }
+            extract_struct_def(
+                edit,
+                enum_ast.syntax(),
+                &variant_name,
+                &field_list.to_string(),
+                start_offset,
+                ctx.frange.file_id,
+                &visibility,
+            );
+            let list_range = field_list.syntax().text_range();
+            update_variant(edit, &variant_name, ctx.frange.file_id, list_range);
+        },
+    );
+}
+
+fn existing_struct_def(db: &RootDatabase, variant_name: &str, variant: &EnumVariant) -> bool {
+    variant
+        .parent_enum(db)
+        .module(db)
+        .scope(db, None)
+        .into_iter()
+        .any(|(name, _)| name.to_string() == variant_name.to_string())
+}
+
+fn insert_import(
+    ctx: &AssistContext,
+    builder: &mut AssistBuilder,
+    path: &ast::PathExpr,
+    module: &Module,
+    module_def: &ModuleDef,
+    path_segment: ast::NameRef,
+) -> Option<()> {
+    let db = ctx.db;
+    let mod_path = module.find_use_path(db, module_def.clone());
+    if let Some(mut mod_path) = mod_path {
+        mod_path.segments.pop();
+        mod_path.segments.push(path_segment.as_name());
+        insert_use_statement(path.syntax(), &mod_path, ctx, builder.text_edit_builder());
+    }
+    Some(())
+}
+
+fn extract_struct_def(
+    edit: &mut AssistDirector,
+    enum_ast: &SyntaxNode,
+    variant_name: &str,
+    variant_list: &str,
+    start_offset: TextSize,
+    file_id: FileId,
+    visibility: &Option<ast::Visibility>,
+) -> Option<()> {
+    let visibility_string = if let Some(visibility) = visibility {
+        format!("{} ", visibility.to_string())
+    } else {
+        "".to_string()
+    };
+    let indent = if let Some(indent) = leading_indent(enum_ast) {
+        indent.to_string()
+    } else {
+        "".to_string()
+    };
+    let mut buf = String::new();
+
+    format_to!(
+        buf,
+        r#"{}struct {}{};
+
+{}"#,
+        visibility_string,
+        variant_name,
```

Comment:
This is actually a `field_list`, right?

Reply:
No, field list is below, I decompose variant into field list and variant name and in field list I add pub modifiers